### ISSUE TITLE
KomodoIDE: In Windows, Python prints in the output box get linux EOL instead of windows EOL

### ISSUE THREAD

**Issue by fjossandon** [Komodo/KomodoEdit#1000]
I'm using KomodoIDE 9.3.1 in Windows 10 to debug a python script (Python 2.7.10 x64), and all "print" or "sys.stdout.write()" generates lines in the Output box that only use LF line ending (linux), instead of the expected CRLF for Windows standard output, which is inconvenient when I just want to copy/paste the output in a notepad. On the other hand, writing to files using the filehandle "write()" function works normal.

This was quite unexpected, since I have been using Komodo for a long time while coding Perl scripts, and Komodo have always used the correct line-ending while printing on Perl.

In summary, for Python:
print "Test" -> LF
sys.stdout.write("Test\n") -> LF
file.write("Test\n") -> CRLF

**Comment by mitchell-as** [comment]
Indeed newline handling between platforms and programming languages can be quite confusing! For consistency, Komodo reads from stdout and stderr of subprocesses (like debugging) in "universal newline" mode. So unless you or the language explicitly sends CRLF, Komodo will read all newlines as simply LF, which appears consistent with what you've observed with Python and Perl.

The only oddity is your file writing example. However, this is explained by your opening of the file in non-binary mode. In that mode, any newlines (LF) are automatically interpreted by Windows as CRLF and written as such.

**Comment by fjossandon** [comment]
But is not consistent at all because both language print different. Perl is printing to the Output box using CRLF, and Python is printing using LF although it should use CRLF in Windows. So why only Python fails to use the correct line ending and only to the Output box but not to the file?

Do you mean that Perl is passing CRLF to Komodo and is Python fault for passing LF instead? When Komodo is running in Windows, shouldn't it add the CR if missing when printing to the standard output?

Running from CMD, the Python prints can be copy/pasted and displayed properly, but running from inside Komodo the CR is missing.

**Comment by mitchell-as** [comment]
Yes, it appears that Perl is explicitly sending CRLF since it detects it is on the Windows platform. (I don't know for sure as I don't know Perl.)

As I mentioned earlier, when you are writing to files (stdout counts as a "file"), you have opened that file in a particular "mode". (Typically binary vs. non-binary). In non-binary mode (which is often the default in any language), Windows likes to handle newlines on its own and thus uses CRLF in place of LF. In binary mode, Windows does not meddle with line endings at all. Komodo's subprocess I/O routines read in binary mode, which does not allow Windows to meddle with line endings at all. Therefore it will read exactly what it was passed.

No, Komodo will not take liberties to add a CR on Windows. It prefers to have all output in LF on any platform, but will not strip CR from CRLF if that was what it was given (i.e. no data loss).

When you run Python from CMD, that stdout is inherently opened (by Windows) in non-binary mode, and thus you get modified CRLF line endings.

I hope this is more clear.

**Comment by fjossandon** [comment]
Yes, I understand now. This means that Perl (or at least Strawberry Perl) adds the CR for both stdout and non-binary files, while Python only adds it for non-binary files.
Thanks.
